Count sigops using generics

upbqdn opened the pull request:
## Motivation

- Close #9669.

## Solution

- Define a new `Sigops` trait.
- Impl the trait for relevant tx types.
- Clean-ups:
	- Use `u32` instead of `u64` for the sigops count since that's what the script interpreter, which returns the counts for tx inputs and outputs, uses.
	- Move tests to their own files.

### PR Checklist

- [x] The PR name is suitable for the release notes.
- [x] The solution is tested.
- [x] The documentation is up to date.

upbqdn committed:
Move script tests to separate tests.rs module

upbqdn committed:
Switch zcash_script to Git dependency

upbqdn committed:
Add Sigops trait for counting transparent signature operations in txs

upbqdn committed:
Implement Sigops trait for Transaction

upbqdn committed:
Implement Sigops for UnminedTx

upbqdn committed:
Implement Sigops for CachedFfiTransaction

upbqdn committed:
Add Sigops impl for zcash_primitives Transaction

upbqdn committed:
Refactor to use Transaction::sigops instead of legacy_sigop_count

upbqdn committed:
Remove legacy_sigop_count function

upbqdn committed:
Remove unused Transaction import

Updated to use fully qualified Transaction paths to resolve ambiguity.
Removed unused Transaction import from module scope.

upbqdn committed:
Use u32 instead of u64 and rename legacy_sigop_count to sigops

upbqdn committed:
Don't skip Canopy in Regtest tests

upbqdn committed:
Don't skip Canopy in Regtest tests

upbqdn committed:
Support Canopy commitment bytes and adjust tests

upbqdn committed:
Support Canopy commitment bytes and adjust tests

upbqdn committed:
Support Canopy commitment bytes and adjust tests

upbqdn committed:
Update zcash_script to 0.3.2 and remove patch for git dependency

upbqdn committed:
Reorder imports

upbqdn committed:
Reorder imports

Standardize import ordering to improve code maintainability and reduce
merge conflicts. No functional changes introduced.

upbqdn commented:
@gustavovalverde, could you take a look at the timeouts? They don't seem related to the PR.

upbqdn commented on zebra-script/src/lib.rs:
```diff
@@ -177,283 +180,63 @@ impl CachedFfiTransaction {
-        } as u64;
+/// Trait for counting the number of transparent signature operations
+/// in the transparent inputs and outputs of a transaction.
+pub trait Sigops {
```
Wouldn't it be better to have this trait in `zcash_script`?

upbqdn replied:
Actually, no, because then I'd have to pull `zcash_script` into `zcash_primitives` to be able to implement the trait.

arya2 commented on zebra-script/src/lib.rs:
```diff
@@ -177,283 +180,63 @@ impl CachedFfiTransaction {
-        verifier.is_valid(0)?;
-
-        Ok(())
+impl Sigops for zcash_primitives::transaction::Transaction {
+    fn scripts(&self) -> impl Iterator<Item = &[u8]> {
+        self.transparent_bundle().into_iter().flat_map(|bundle| {
+            (!bundle.is_coinbase())
+                .then(|| bundle.vin.iter().map(|i| i.script_sig.0.as_slice()))
+                .into_iter()
+                .flatten()
+                .chain(bundle.vout.iter().map(|o| o.script_pubkey.0.as_slice()))
+        })
```
Is this being tested or used anywhere?

upbqdn replied:
I use it three branches up the stack of the shielded coinbase saga. :D I'll try thinking of some vectors for tests.

upbqdn replied:
I think tests will be much easier to add once I have the handling of the tx builder done, so I can then use it to create txs for which I'll know the sigops count.

upbqdn replied:
The main purpose of this PR is to have sigops counting ready for `zcash_primitives::transaction::Transaction` so that the subsequent PRs are easier to review.

arya2 reviewed (dismissed):
Look good to me as long as there's a test for the `Sigops` impl for the `zcash_primitives` transaction type.

It would be nice to have `Sigops` implemented in `zcash-script`, but not necessary.

I particularly like the use of `try_fold()`.

conradoplg commented on zebra-script/Cargo.toml:
```diff
@@ -23,7 +23,9 @@ comparison-interpreter = []
-zebra-chain = { path = "../zebra-chain", version = "1.0.0" }
+zcash_primitives = { workspace = true }
+
+zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.46" }
```
```suggestion
zcash_primitives = { workspace = true }
zebra-chain = { path = "../zebra-chain", version = "1.0.0" }
```

(or `workspace = true` for it too?)

upbqdn replied:
> (or `workspace = true` for it too?)

That doesn't work.

upbqdn replied:
Commited in [0bdd5b2](https://github.com/ZcashFoundation/zebra/pull/9670/commits/0bdd5b28e90458f1f4f6991d3ce4672b16fb5917).

conradoplg commented on zebra-script/src/lib.rs:
```diff
@@ -177,283 +180,63 @@ impl CachedFfiTransaction {
+    /// Returns the number of transparent signature operations in the
+    /// transparent inputs and outputs of the given transaction.
+    fn sigops(&self) -> Result<u32, zcash_script::Error> {
+        let interpreter = get_interpreter(
```
```suggestion
        // Create a dummy interpreter since these inputs are not used to count
        // the sigops
        let interpreter = get_interpreter(
```

upbqdn replied:
What do "these inputs" refer to?

conradoplg replied:
The inputs to `get_interpreter` - feel free to improve the message, I just wanted to keep the spirit of the original comment (explain why it's OK to use dummy parameters) which got lost in the refactoring

upbqdn replied:
Ah, you don't mean "tx inputs" but the fn args. It took me a while. :D

upbqdn replied:
Addressed in [Clarify use of dummy values in sigops counting](https://github.com/ZcashFoundation/zebra/pull/9670/commits/39cdd625b7918269544d2baf808274dca5c7cee7).

conradoplg commented on zebra-script/src/lib.rs:
```diff
@@ -177,283 +180,63 @@ impl CachedFfiTransaction {
+            .iter()
+            .map(|input| match input {
+                transparent::Input::PrevOut { unlock_script, .. } => unlock_script.as_raw_bytes(),
+                transparent::Input::Coinbase { .. } => &[],
```
I'm a bit worried about passing an empty slice to `legacy_sigop_count_script()` which goes all the way down to C++ code. It looks like it works, but I wonder if it's safer to skip those by using `filter_map()` instead of `map()`

upbqdn replied:
Done in [Filter coinbase inputs from script iterator in Sigops implementation](https://github.com/ZcashFoundation/zebra/pull/9670/commits/11dd9b1e69090d18122ab3c255eb763f6db0d8fb).

conradoplg commented on zebrad/tests/acceptance.rs:
```diff
@@ -2973,8 +2973,8 @@ async fn trusted_chain_sync_handles_forks_correctly() -> Result<()> {
 
-    let network = Network::new_regtest(Default::default());
-    let mut config = os_assigned_rpc_port_config(false, &network)?;
+    let net = Network::new_regtest(Default::default());
+    let mut config = os_assigned_rpc_port_config(false, &net)?;
```
I don't mind leaving this here for now but I think we should decide as a team on whether we start renaming `network` to `net` over the code base

upbqdn replied:
I implicitly thought it would be a good idea. I'm happy to discuss.

conradoplg reviewed (commented):
Looks good, left some fixes and suggestions

conradoplg reviewed (changes requested):
(blocking on the import fix)

upbqdn committed:
Merge branch 'main' into refactor-sigops

upbqdn committed:
Update zebra-script/Cargo.toml

Co-authored-by: Conrado Gouvea <conrado@zfnd.org>

upbqdn committed:
Merge branch 'refactor-sigops' of github.com:ZcashFoundation/zebra

upbqdn committed:
Clarify use of dummy values in sigops counting

upbqdn committed:
Filter coinbase inputs from script iterator in Sigops implementation

Previously, coinbase inputs yielded empty byte slices; now they are
excluded. This simplifies downstream consumers and avoids passing empty
scripts.
